### PR TITLE
Allow to use star as repository name in imagestreamimport

### PR DESCRIPTION
**Pull request by legionus**
Fix https://github.com/openshift/origin/issues/11773

@smarterclayton @liggitt @miminar @mfojtik Please review.

### CODE REVIEW

File: pkg/image/api/validation/validation.go [hunk 1]
```diff
@@ -308,6 +308,9 @@ func ValidateImageStreamImport(isi *api.ImageStreamImport) field.ErrorList {
 			if len(spec.From.Name) == 0 {
 				errs = append(errs, field.Required(imagesPath.Index(i).Child("from", "name"), ""))
 			} else {
+				if spec.From.Name == "*" {
```
Comment (liggitt) [inline pkg/image/api/validation/validation.go]: I wasn't expecting this to be allowed in the API, only used client side. What will allowing this in the API do?

Comment (legionus) [inline pkg/image/api/validation/validation.go]: https://github.com/openshift/origin/pull/11440#issuecomment-258346246

Comment (liggitt) [inline pkg/image/api/validation/validation.go]: ah. a comment here explaining why we do this would be good.

**Comment by legionus** [comment]
[test]

File: pkg/image/api/validation/validation.go (hunk 1 above)
Comment (legionus) [inline pkg/image/api/validation/validation.go]: The problem is that `ParseDockerImageReference` the `*` is no longer valid name. Clients use this character to look up Docker images.

**Comment by mfojtik** [comment]
@smarterclayton can you please review/merge? This is 1.4 blocker.

File: pkg/image/api/validation/validation.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/image/api/validation/validation.go]: Add a comment here to indicate why this exists (legacy client sent this value)

File: pkg/image/importer/importer.go [hunk 2]
```diff
@@ -110,10 +110,18 @@ func (i *ImageStreamImporter) importImages(ctx gocontext.Context, retriever Repo
...
 		}
-		ref, err := api.ParseDockerImageReference(from.Name)
-		if err != nil {
-			isi.Status.Images[i].Status = invalidStatus("", field.Invalid(field.NewPath("from", "name"), from.Name, fmt.Sprintf("invalid name: %v", err)))
-			continue
+		var (
+			err error
+			ref api.DockerImageReference
+		)
+		if from.Name != "*" {
```
Comment (smarterclayton) [inline pkg/image/importer/importer.go]: If you get this you can just skip it.  DockerImage = "*" returns no result.  So up with `from.Kind != "" || from.Name == "*"`

Comment (legionus) [inline pkg/image/importer/importer.go]: @smarterclayton I already tried this https://gist.github.com/legionus/13042d3682fbba1ca6db53b7c5b1e653 and got:
```
$ oc new-app -L 
error: can't look up Docker image "*":
```
with my changes this command returns:
```
Templates (oc new-app --template=<template>)
-----
cakephp-mysql-example
  Project: openshift
  An example CakePHP application with a MySQL database
dancer-mysql-example
  Project: openshift
  An example Dancer application with a MySQL database
django-psql-example
  Project: openshift
  An example Django application with a PostgreSQL database
jenkins-ephemeral
  Project: openshift
  Jenkins service, without persistent storage.
WARNING: Any data stored will be lost upon pod destruction. Only use this template for testing
jenkins-persistent
  Project: openshift
  Jenkins service, with persistent storage.
You must have persistent volumes available in your cluster to use this template.
jenkins-pipeline-example
  Project: openshift
  This example showcases the new Jenkins Pipeline integration in OpenShift, which performs continuous integration and deployment right on the platform. The template contains a Jenkinsfile - a definition of a multi-stage CI process - that leverages the underlying OpenShift platform for dynamic and scalable builds. OpenShift integrates the status of your pipeline builds into the web console allowing you to see your entire application lifecycle in a single view.
mariadb-ephemeral
  Project: openshift
  MariaDB database service, without persistent storage. WARNING: Any data stored will be lost upon pod destruction. Only use this template for testing
mongodb-ephemeral
  Project: openshift
  MongoDB database service, without persistent storage. WARNING: Any data stored will be lost upon pod destruction. Only use this template for testing
mysql-ephemeral
  Project: openshift
  MySQL database service, without persistent storage. WARNING: Any data stored will be lost upon pod destruction. Only use this template for testing
nodejs-mongodb-example
  Project: openshift
  An example Node.js application with a MongoDB database
postgresql-ephemeral
  Project: openshift
  PostgreSQL database service, without persistent storage. WARNING: Any data stored will be lost upon pod destruction. Only use this template for testing
rails-postgresql-example
  Project: openshift
  An example Rails application with a PostgreSQL database

Image streams (oc new-app --image-stream=<image-stream> [--code=<source>])
-----
jenkins
  Project: openshift
  Tags:    1, latest
mariadb
  Project: openshift
  Tags:    10.1, latest
mongodb
  Project: openshift
  Tags:    2.4, 2.6, 3.2, latest
mysql
  Project: openshift
  Tags:    5.5, 5.6, latest
nodejs
  Project: openshift
  Tags:    0.10, 4, latest
perl
  Project: openshift
  Tags:    5.16, 5.20, latest
php
  Project: openshift
  Tags:    5.5, 5.6, latest
postgresql
  Project: openshift
  Tags:    9.2, 9.4, 9.5, latest
python
  Project: openshift
  Tags:    2.7, 3.3, 3.4, 3.5, latest
ruby
  Project: openshift
  Tags:    2.0, 2.2, 2.3, latest
wildfly
  Project: openshift
  Tags:    10.0, 10.1, 8.1, 9.0, latest
```

File: pkg/image/api/validation/validation.go (hunk 1 above)
Comment (legionus) [inline pkg/image/api/validation/validation.go]: @smarterclayton I added the comment.

**Comment by legionus** [comment]
[test]

**Comment by legionus** [comment]
Flake https://github.com/openshift/origin/issues/11777

**Comment by legionus** [comment]
[test]

**Comment by legionus** [comment]
Again. Flake #11777. [test]

File: pkg/image/importer/importer.go [hunk 3]
```diff
@@ -217,11 +225,20 @@ func (i *ImageStreamImporter) importFromRepository(ctx gocontext.Context, retrie
 	if from.Kind != "DockerImage" {
 		return
 	}
-	ref, err := api.ParseDockerImageReference(from.Name)
-	if err != nil {
-		status.Status = invalidStatus("", field.Invalid(field.NewPath("from", "name"), from.Name, fmt.Sprintf("invalid name: %v", err)))
-		return
+	var (
```
Comment (smarterclayton) [inline pkg/image/importer/importer.go]: Add a todo that this should be removed in 1.6

Comment (legionus) [inline pkg/image/importer/importer.go]: @smarterclayton OK, I will. But our client still [sends](https://github.com/openshift/origin/blob/master/pkg/generate/app/dockerimagelookup.go#L197) this symbol to see the list.
why this workaround should be removed ?

**Comment by smarterclayton** [comment]
Just that the workaround is time limited and we will eventually remove
it (in a future release) so that complexity is reduced.  We'll
announce it's deprecated in 1.4 (you should add a release note if you
havent)

**Comment by legionus** [comment]
@smarterclayton I added a comment and added a link to your comment so it's not lost.

**Comment by smarterclayton** [comment]
Lgtm [merge] thanks

**Comment by mfojtik** [comment]
flake?: https://github.com/openshift/origin/issues/11867

[merge]

**Pull request merged**